src/DelegateDecompiler/MethodBodyDecompiler.cs

```diff
@@ -17,35 +17,66 @@ public static LambdaExpression Decompile(MethodInfo method)
         public static LambdaExpression Decompile(MethodInfo method, Type declaringType)
         {
             var args = method.GetParameters()
-                .Select(p => (Address) Expression.Parameter(p.ParameterType, p.Name))
+                .Select(p => (Address)Expression.Parameter(p.ParameterType, p.Name))
                 .ToList();
 
             var methodType = declaringType ?? method.DeclaringType;
             if (!method.IsStatic)
                 args.Insert(0, Expression.Parameter(methodType, "this"));
 
-            var expression = method.IsVirtual
+            var expression = method.IsVirtual && !DecompileExtensions.ConcreteCalls.ContainsKey(new Tuple<object, MethodInfo>(args.FirstOrDefault(), method))
                 ? DecompileVirtual(methodType, method, args)
                 : DecompileConcrete(method, args);
 
             var optimizedExpression = new OptimizeExpressionVisitor().Visit(expression);
 
-            return Expression.Lambda(optimizedExpression, args.Select(x => (ParameterExpression) x.Expression));
+            return Expression.Lambda(optimizedExpression, args.Select(x => (ParameterExpression)x.Expression));
         }
 
-        static Expression DecompileConcrete(MethodInfo method, IList<Address> args)
+        internal static Expression DecompileConcrete(MethodInfo method, IList<Address> args)
         {
+            HashSet<Tuple<object, MethodInfo>> baseCalls = new HashSet<Tuple<object, MethodInfo>>();
+
             var body = method.GetMethodBody();
             var addresses = new VariableInfo[body.LocalVariables.Count];
             for (var i = 0; i < addresses.Length; i++)
                 addresses[i] = new VariableInfo(body.LocalVariables[i].LocalType);
             var locals = addresses.ToArray();
 
+            if (!method.IsStatic && DelegateDecompiler.DecompileExtensions.ConcreteCalls.ContainsKey(new Tuple<object, MethodInfo>(args.First().DiscardConversion(), method)))
```

Comment:
Once a concrete implementation is to be decompiled we register any possible call to overridden member to be decompiled concretely for this particular instance.